### ISSUE TITLE
Change in import behavior for spaces

### ISSUE THREAD

**Issue by Necr0x0Der** [trueagi-io/hyperon-experimental#592]
With the new module system, I got the following error:
```
(Error (import! GroundingSpace-0x12a7846d8 space.metta) import! destination argument must be a symbol atom naming a new space, or &self.  Found: GroundingSpace-0x12a7846d8)
```
Why can't we import into the already existing space? It is inconvenient.

**Comment by luketpeterson** [comment]
The short answer is because the behavior of the `import!` atom is a bad conceptual fit for the abstractions in the new MeTTa (or any language's) module system.  For example, it wouldn't make any sense in Python to say `from numpy import * into torch`.  If torch needed something from numpy it would do the import itself.

In MeTTa there is also a question about what to do about tokens from the modules, although that question also exists for just importing into a new space.  https://github.com/trueagi-io/hyperon-experimental/blob/8540b90d8669484e945fb78d0d11135b68de9233/docs/modules_internal_discussion.md?plain=1#L41C1-L42C1

The arguments to `import!` need to be redone to better reflect the underlying semantics of the operation as discussed here: https://github.com/trueagi-io/hyperon-experimental/issues/509 and also https://github.com/trueagi-io/hyperon-experimental/pull/580#discussion_r1491332304

But this issue seems to be about providing a convenience.  So I have two ideas:
1.) The `inline!` operation.  I could add a space argument to it.  So any atoms created by the inlined code would go into the specified space. `inline!` seems to be what a lot of people want for a lot of reasons.
2.) An automatic space-merge if the specified space already exists.  I don't like this because it further muddies up what `import!` is conceptually supposed to do.  But it would bring back the old behavior.

**Comment by Necr0x0Der** [comment]
> The current behavior of the import! operation is a bad conceptual fit for the abstractions in the new MeTTa (or any language's) module system. For example, it wouldn't make any sense in Python to say from numpy import * into torch

I would repeat that the initial proposal way to consider importing modules and loading spaces as two different use cases. And initial implementation of `import!` was more about loading spaces than about importing modules. But `import!` and `extend-py!` was merged into one, instead of separating them further. However, I disagree with `from numpy import * into torch` analogy. If you keep importing modules as spaces and since MeTTa does have a good and clear conceptual explanation what it means to load atoms to spaces, and in contrast to many languages it allows working directly with code and it is extremely useful to work (add / remove / etc.) with code in separate spaces, removing the possibility to load something into a space, was definitely a bad idea. Even if we restore this functionality, saying that it's a bad conceptual fit to have the possibility to load additional content into modules at runtime is debatable. Why a human can add something to a module manually, but MeTTa cannot? I believe not all the design patterns of "all other languages" are applicable to MeTTa because of its self-reflective nature.

I don't like `inline!` name too much. In any case, I wonder, if loading another space into an imported module space will work.

**Comment by Necr0x0Der** [comment]
All in all, I disagree with the "conceptual" argument. If we can import something into `&self` space, we should be able to import something in another space, e.g. via executing `import!` to `&self` in the context of this space. Making artificial restriction to this is just wrong. If the restriction is made not because of conceptual reasons, but because of natural implementation, this is another argument, which has more weight.

**Comment by luketpeterson** [comment]
> All in all, I disagree with the "conceptual" argument. If we can import something into &self space, we should be able to import something in another space

That's precisely the issue.  `import! &self` is just a legacy way to trigger "import *" behavior, but it doesn't make sense to think of `import!` as having a destination.  All import behaviors ("import module as name", "import * from module", "import foo from module [as name]") effectively import into the currently running module.  ie. `self`.  For example `!(import! &new-space my_mod)` imports a new token into the `self` module's Tokenizer, to reference the sub-module's space.

> Programming language design is mostly about convenience.

100% agreed on that point.  So we (our team) need to do the hard work of thinking through all the edge cases, so we can come up with an API and a language that presents the best balance of an ergonomic, intuitive, and coherent interface with all the power and expressiveness people will want.

So along those lines, what convenience do you think makes sense to offer?

One option is a merge-space operator, so you could run
```
!(import! &new_space the_mod)
!(merge-space &new_space &destination)
```
This is more verbose than `!(import! &destination the_mod)`, but `merge-space` will certainly have uses beyond just importing.  Also another downside of this pattern is that `&new_space` will be unnecessary leftover trash.

An alternative idea might be to not use `import!` here, and instead add something to provide an accessor for a module's space.  Like this:
```
!(merge-space (mod-space! the_mod) &destination)
```

What do you think?

> I don't like `inline!` name too much.

Is it just the name you don't like, or the semantic as well?

**Comment by Necr0x0Der** [comment]
> One option is a merge-space operator, so you could run...

I'd propose to start with use-cases, and make it convinient for them. My current use case was that there is a bunch of metta-files constituting bio-atomspace, and all of them are supposed to be loaded into the same space. It will be inconvenient to load them first in their own spaces (and assigning tokens to them), and then to merge them. I was also thinking about the function which would return the space atom, into which the content is loaded (unlike `import!`). However, it's still not precisely what needed, because loading additional content to the existing space might be more efficient than loading it into a separate space and then merging them (it, of course, depends on how spaces are implemented; if we have an algebra over spaces, then merge of spaces should not necesserily be a "physical" merge; OTOH such algebra poses some challenging for indexing mechanisms, so I wouldn't make it obligatory for any kind of spaces). In any case, while merging spaces can be useful by itself for other cases, I wouldn't introduce it for this case. 
Thus, I think we need to restore the former functionality of loading additional content into the existing spaces. Obviously, it should not necessarily be `import!`. Although if we go away from importing to anywhere but `&self`, then we need to make it less misleading, because `&self` is a space. And it is definitely an inconsistent notation that we can pass `&self` space to `import!` but we cannot pass another space, and instead we need to specify a token for a new space. This is a separate discussion, however, and I wouldn't like to introduce rush changes to `import!` syntax without really thinking through it carefully and coming up with the version, which will be stable. Right now, it's definitely not a blocking issue. What is more urgent is the now missing functionality.
OK, let it not be `import!`. We need to load the content of a file to the already existing space. But there is still more that one sub-cases. One may want to load a pure space (but it can possibly contain numbers and strings as grounded atoms), or one can want to interpret the code of another script in the content of the existing space. This situation appeared in metta-motto, and after new modules were introduced I had to emulate this behavior in the metta-motto itself (which is ok for better controllability). As for bio-atomspace case, it will work for me with this PR https://github.com/trueagi-io/hyperon-experimental/pull/593 (which is needed not only to workaround the changed behavior, but for performance reasons as well).
Thus, I'd wait for other use cases to understand if we need first loading a pure spac, or running a script in context of existing module/space. Meanwhile, we have other useful features to implement like module downloading and installation.

> Is it just the name you don't like, or the semantic as well?

It's the name. Although the details of semantics are to be discussed.

**Comment by vsbogd** [comment]
Can we say that we have "importing" and "loading" operations. When module is imported then atoms of the module become a part of a destination module and tokens of the module are used to parse following atoms of the destination module. (Strictly speaking `import!` not necessary loads atoms into the destination module but it makes them available for the destination module. But at the moment it is implemented by adding atoms into the destination module.) When module is loaded then atoms of the module also become a part of destination space but tokens of the module are used only to parse the modules itself. Thus from the atoms perspective importing and loading are not different at least before we have "import as" which will extend the semantics of the import. From this perspective `inline!` not so bad name but `load!` could be more clear. And `import!` is different because it treats tokens (and potentially atoms) differently.

**Comment by Necr0x0Der** [comment]
@luketpeterson , it should be prioritized as many metta coders complain about this

**Comment by TeamSPoon** [comment]
I like the name `include!` because I use it to include functions defined in  .metta files.  Also there are cases when I am including the same files that have directives I need into multiple metta files.

**Comment by luketpeterson** [comment]
I realized (after implementing it) that a `merge-atoms` operation to copy all atoms from a `src` space to a `dst` space is unneeded on account of the nested-space behavior.  Therefore: `!(add-atom &dst_space (mod-space! mod_name))` should be enough.

**Comment by luketpeterson** [comment]
mod-space!  op atom added in https://github.com/trueagi-io/hyperon-experimental/pull/615

**Comment by luketpeterson** [comment]
Assigning back to issue originator to confirm change addresses the root issue.

**Comment by Necr0x0Der** [comment]
@luketpeterson , what will be the canonical way to load several scripts into the same space?

**Comment by luketpeterson** [comment]
> @luketpeterson , what will be the canonical way to load several scripts into the same space?

When you say "canonical" what do you mean?  This will work if the modules are findable by the module resolution logic:
```
!(add-atom &the_space (mod-space! mod1))
!(add-atom &the_space (mod-space! mod2))
!(add-atom &the_space (mod-space! mod3))
```

If they're not findable, you can do:
```
!(register-module! /path/to/script1.metta)
!(add-atom &the_space (mod-space! script1))
!(register-module! /path/to/script2.metta)
!(add-atom &the_space (mod-space! script2))
```

**Comment by Necr0x0Der** [comment]
I mean, previously, we could do:
```
!(bind! &space (new-space))
!(import! &space script1)
!(import! &space script2)
```

What is the equivalent now? Or are we waiting for `inline!` from you?

**Comment by luketpeterson** [comment]
If `script1` and `script2` are findable, the exact same behavior will be accomplished by:
```
!(bind! &space (new-space))
!(add-atom &space (mod-space! script1))
!(add-atom &space (mod-space! script2))
```

**Comment by Necr0x0Der** [comment]
> the same behavior will be accomplished by:

It doesn't seem to work. I'm still getting the error `(Error (import! GroundingSpace-0x1035d4ba8 a1_symbols) import! destination argument must be a symbol atom naming a new space, or &self.  Found: GroundingSpace-0x1035d4ba8)`

**Comment by Necr0x0Der** [comment]
It doesn't work as desired.
```
!(bind! &space (new-space))
!(add-atom &space (mod-space! a1_symbols))

!(match &space $x $x)
```
returns `GroundingSpace-top:a1_symbols` and doesn't delegate the matching process to the inserted `a1_symbols` space.

**Comment by luketpeterson** [comment]
Huh.  I'm confused because `import!` has created nested sub-spaces from the beginning.  https://github.com/trueagi-io/hyperon-experimental/blob/586e2b2a050fe79bd7b7edb7055e5d2593730200/lib/src/metta/runner/stdlib.rs#L117

I am not sure why the nested sub-space are not traversed in this case.  I'll need to investigate.

**Comment by luketpeterson** [comment]
Investigating further, it seems to be a quoting / reducing issue.
```
!(bind! &space (new-space))
!(bind! &a1_space (mod-space! a1_symbols))
!(add-atom &space &a1_space)
!(match &space (isa red $x) $x)
```
produces `[color]`, but
```
!(bind! &space (new-space))
!(add-atom &space (mod-space! a1_symbols))
!(match &space (isa red $x) $x)
```
produces `[]`

**Comment by luketpeterson** [comment]
Indeed.  `!(add-atom &self (mod-space! a1_symbols))` leads to the unreduced expr `(mod-space! a1_symbols)` being added to the space.  But `get-atoms` was showing me the same results in both paths because it was reducing the output.  I had to inspect the space internals to see they contained different atoms.

**Comment by vsbogd** [comment]
We can do:
```
(= (inline! $dst $src)
  (let $x (mod-space! $src) (add-atom $dst $x)))
!(inline! &space a1_symbols)
```
But such `inline!` will not insert atoms from `a1_symbols` to `&space`.

**Comment by luketpeterson** [comment]
Hard to say what the most user-friendly option would be.  I guess my preference would be for this:
```
(= (merge $dst $space)  (let $x $space (add-atom $dst $x)))
```
which could be called as
```
!(merge &space (mod-space! src_mod))
```

**Comment by luketpeterson** [comment]
I raised https://github.com/trueagi-io/hyperon-experimental/pull/617 to add the `merge-space` op in MeTTa code.

```
!(bind! &space (new-space))
!(merge-space &space (mod-space! a1_symbols))

!(match &space (isa red $x) $x)
```
The above code now returns `[color]`

**Comment by luketpeterson** [comment]
@Necr0x0Der Does the `add_reduct` operation, as demonstrated above, solve this issue?

**Comment by Necr0x0Der** [comment]
Yes, it works as a workaround. However, it looks a little bit indirect, so if we suppose it to be an idiomatic way to achieve this functionality, we may want to add it into standard examples. Maybe, we need a sort of short tutorial on imports covering different cases.

**Comment by patham9** [comment]
I haven't followed the changes to the import mechanism, except of the unification of `extend-py!` and `import!` which, albeit I personally don't like two quite different things being sold under the same umbrella, I at the same time find it convenient (it makes it simpler for the user), and can hence agree with the design decision.

Now when it comes to complicating the idea of importing spaces into other spaces, I don't see why one would want to add additional complexity for the user when previously an import! was all that was necessary. MeTTa has very different requirements in regards how spaces and code can be modified at runtime than Python.
